lib/templates.js

```diff
@@ -184,15 +184,59 @@ const createGroupsTemplateForService = (serviceName, serviceDefinition, appConfi
 
 /**
  * This is a template function to create EntityType object for EntityTypes Array.
+ * Ensures correct visibility assignment based on referenced resources.
+ *
  * @param { object } appConfig  The configuration object.
- * @param { Array } packageIds The identifiers of packages.
- * @param { object } entity
+ * @param { Array } packageIds The available package identifiers.
+ * @param { object } entity The entity definition.
  * @returns { object } An object for the EntityType.
  */
 const createEntityTypeTemplate = (appConfig, packageIds, entity) => {
     const ordExtensions = readORDExtensions(entity);
+    const visibility = ordExtensions.visibility || RESOURCE_VISIBILITY.public;
+
+    // Check whether the EntityType is referenced by a restrictive API
+    const referencedByRestrictedAPI = (Array.isArray(appConfig.apiResources) ? appConfig.apiResources : []).some(api =>
+        api.entityTypeMappings?.some(mapping =>
+            mapping.entityTypeTargets.some(target => target.ordId === entity.ordId)
+        ) && [RESOURCE_VISIBILITY.private, RESOURCE_VISIBILITY.internal].includes(api.visibility)
+    );
```

Comment:
Please generalize this check across APIs, Events and Data Products (you can pass a getEntity function to get the difference between resources with entityTypeMappings and plain entityType).